### DISCUSSION TITLE
Conditional rendering is not working for accordion tabs

### DISCUSSION THREAD

**Discussion by adityapatil9191** [data-driven-forms/react-forms#1459] [Need help] [answered]
**Scope**: <DESCRIBE SCOPE OF THIS ISSUE, i.e.: PF4 mapper, Form renderer>
https://codesandbox.io/p/sandbox/react-data-driven-autocomplete-forked-t8x4m5?file=%2Fsrc%2Fcomponents%2Fschema.js%3A114%2C11
**Description**

The tab component should be visible only if form component values present in another section match a certain condition. However that is not working. Please find the schema below for your reference.

**Schema**

const schema = {
  fields: [
    {
      component: "tabs",
      name: "tabs",
      fields: [
        {
          name: "A",
          title: "TAB 1",
          component: "tabs",
          fields: [
            {
              component: "autocomplete",
              name: "a.pr",
              label: "a.pr",
              initialValue: {
                value: "A",
                label: "A",
              },
              options: [
                {
                  value: "A",
                  label: "A",
                },
                {
                  value: "B",
                  label: "B",
                },
                {
                  value: "C",
                  label: "C",
                },
              ],
            },
            {
              component: "autocomplete",
              name: "in.cp",
              label: "in.cp",
              initialValue: {
                value: "A",
                label: "A",
              },
              options: [
                {
                  value: "A",
                  label: "A",
                },
                {
                  value: "B",
                  label: "B",
                },
                {
                  value: "C",
                  label: "C",
                },
              ],
              clearOnUnmount: true,
            },
          ],
        },
        {
          name: "B",
          title: "TAB 2",
          component: "tabs",
          fields: [
            {
              component: "autocomplete",
              name: "a.pr2",
              label: "a.pr2",
              initialValue: {
                value: "A",
                label: "A",
              },
              options: [
                {
                  value: "A",
                  label: "A",
                },
                {
                  value: "B",
                  label: "B",
                },
                {
                  value: "C",
                  label: "C",
                },
              ],
            },
            {
              component: "autocomplete",
              name: "in.cp2",
              label: "in.cp2",
              initialValue: {
                value: "A",
                label: "A",
              },
              options: [
                {
                  value: "A",
                  label: "A",
                },
                {
                  value: "B",
                  label: "B",
                },
                {
                  value: "C",
                  label: "C",
                },
              ],
              clearOnUnmount: true,
            },
          ],
        },
        {
          name: "C",
          title: "TAB 3",
          component: "tabs",
          fields: [
            {
              component: "autocomplete",
              name: "a.pr3",
              label: "a.pr3",
              initialValue: {
                value: "A",
                label: "A",
              },
              options: [
                {
                  value: "A",
                  label: "A",
                },
                {
                  value: "B",
                  label: "B",
                },
                {
                  value: "C",
                  label: "C",
                },
              ],
            },
            {
              component: "autocomplete",
              name: "in.cp3",
              label: "in.cp3",
              initialValue: {
                value: "A",
                label: "A",
              },
              options: [
                {
                  value: "A",
                  label: "A",
                },
                {
                  value: "B",
                  label: "B",
                },
                {
                  value: "C",
                  label: "C",
                },
              ],
              clearOnUnmount: true,
            },
            {
              component: "fieldArray",
              name: "a.sa",
              fieldKey: "field_array",
              title: "fieldArray-1",
              fields: [
                {
                  component: "textField",
                  name: "key",
                  label: "Annotation Name",
                },
                {
                  component: "textField",
                  name: "value",
                  label: "Annotation Value",
                },
              ],
              condition: {
                sequence: [
                  {
                    and: [
                      {
                        when: "a.pr.value",
                        is: "B",
                        notMatch: true,
                      },
                      { when: "in.cp3.value", is: "B" },
                    ],
                    then: {
                      set: {
                        "a.pr3": { value: "C", label: "C" },
                        "a.sa": [
                          {
                            key: "type",
                            value: "nlb",
                          },
                        ],
                      },
                    },
                    else: { visible: true },
                  },
                ],
              },
            },
            {
              component: "textField",
              name: "in.cp4",
              initialValue: "initial",
              clearOnUnmount: true,
              hideField: false,
              condition: {
                when: "in.cp3.value",
                is: "B",
                then: { set: { "in.cp4": "set" } },
              },
            },
          ],
          condition: { when: "a.pr", is: "A", then: { visible: true } },
        },
      ],
    },
  ],
};

export default schema;

**Comment by adityapatil9191** [comment]
Please help / respond

**Comment by rvsia** [comment]
Hey, if I understand your issue correctly, then I think the issue is here:

<img width="630" alt="Snímek obrazovky 2024-05-10 v 7 29 11" src="https://github.com/data-driven-forms/react-forms/assets/32869456/02fe5824-2d11-4aee-ba9e-c4d1e540d5c0">

When you render tab components, there is actually no data driven from component (so no condition is calculated), but you render directly children of the component. You can wrap this section to a new component in the mapper and then just do `formOptions.renderForm(fields)`

**Comment by adityapatil9191** [comment]
@rvsia could you please help me with the code example?

**Reply by adityapatil9191** [reply]
I have mapped the tabs to the tabs component in the schema and I have also added schema for your reference. I would like the condition applied on the tab section mapped to tabs component work. Please find the reference schema changes. I have added the condition below "condition: { when: "a.pr", is: "A", then: { visible: true } }," to the schema but unfortunately it is not working :-(.
{
          name: "C",
          title: "TAB 3",
          component: "tabs",
          fields: [
            {
              component: "autocomplete",
              name: "a.pr3",
              label: "a.pr3",
              initialValue: {
                value: "A",
                label: "A",
              },
              options: [
                {
                  value: "A",
                  label: "A",
                },
                {
                  value: "B",
                  label: "B",
                },
                {
                  value: "C",
                  label: "C",
                },
              ],
            },
            {
              component: "autocomplete",
              name: "in.cp3",
              label: "in.cp3",
              initialValue: {
                value: "A",
                label: "A",
              },
              options: [
                {
                  value: "A",
                  label: "A",
                },
                {
                  value: "B",
                  label: "B",
                },
                {
                  value: "C",
                  label: "C",
                },
              ],
              clearOnUnmount: true,
            },
            {
              component: "fieldArray",
              name: "a.sa",
              fieldKey: "field_array",
              title: "fieldArray-1",
              fields: [
                {
                  component: "textField",
                  name: "key",
                  label: "Annotation Name",
                },
                {
                  component: "textField",
                  name: "value",
                  label: "Annotation Value",
                },
              ],
              condition: {
                sequence: [
                  {
                    and: [
                      {
                        when: "a.pr.value",
                        is: "B",
                        notMatch: true,
                      },
                      { when: "in.cp3.value", is: "B" },
                    ],
                    then: {
                      set: {
                        "a.pr3": { value: "C", label: "C" },
                        "a.sa": [
                          {
                            key: "type",
                            value: "nlb",
                          },
                        ],
                      },
                    },
                    else: { visible: true },
                  },
                ],
              },
            },
            {
              component: "textField",
              name: "in.cp4",
              initialValue: "initial",
              clearOnUnmount: true,
              hideField: false,
              condition: {
                when: "in.cp3.value",
                is: "B",
                then: { set: { "in.cp4": "set" } },
              },
            },
          ],
          condition: { when: "a.pr", is: "A", then: { visible: true } },
        },

**Reply by rvsia** [reply]
Can you please share a new codesandbox?

**Reply by adityapatil9191** [reply]
@rvsia Please find the new sandbox for reference. https://codesandbox.io/p/sandbox/react-data-driven-autocomplete-forked-t8x4m5?file=%2Fsrc%2Fcomponents%2Fschema.js

**Reply by rvsia** [reply]
I am getting "Sandbox not found". :-/

**Reply by adityapatil9191** [reply]
Apologies for the inconvenience caused.  I have updated the permissions now. [1](https://codesandbox.io/p/sandbox/react-data-driven-autocomplete-forked-t8x4m5?file=%2Fsrc%2Fcomponents%2Fschema.js%3A114%2C11&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clv0oza0g00073b6lls6x0vih%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clv0oza0f00033b6lvh86ct40%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clv0oza0f00043b6ly30y4vfo%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clv0oza0f00063b6lpvn9mjyf%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B63.489777278156396%252C36.510222721843604%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clv0oza0f00033b6lvh86ct40%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clv0oza0f00023b6l7zowrkyz%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Fcomponents%252FcomponentMapper.jsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clw0nyo8l00023b6lmv6omwgk%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clw0p03nq00023b6ln0vyobx7%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A114%252C%2522startColumn%2522%253A11%252C%2522endLineNumber%2522%253A114%252C%2522endColumn%2522%253A11%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252Fcomponents%252Fschema.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clv0oza0f00033b6lvh86ct40%2522%252C%2522activeTabId%2522%253A%2522clw0p03nq00023b6ln0vyobx7%2522%257D%252C%2522clv0oza0f00063b6lpvn9mjyf%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clv0oza0f00053b6lomae1dno%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clv0oza0f00063b6lpvn9mjyf%2522%252C%2522activeTabId%2522%253A%2522clv0oza0f00053b6lomae1dno%2522%257D%252C%2522clv0oza0f00043b6ly30y4vfo%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clv0oza0f00043b6ly30y4vfo%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Reply by rvsia** [reply, marked as answer]
@adityapatil9191 Here is fixed version [[1](https://codesandbox.io/p/sandbox/react-data-driven-autocomplete-forked-gscyhs?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clw0p75sm00073p6g3rsytx0o%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clw0p75sm00033p6gdkgcil0f%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clw0p75sm00043p6ggf1568jt%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clw0p75sm00063p6gbgvq5cit%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clw0p75sm00033p6gdkgcil0f%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clw0p75sm00023p6gnduzhleu%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Fcomponents%252Fschema.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clw0p8510004c3p6gxo3ivu91%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A144%252C%2522startColumn%2522%253A3%252C%2522endLineNumber%2522%253A144%252C%2522endColumn%2522%253A3%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252Fcomponents%252FcomponentMapper.jsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clw0p75sm00033p6gdkgcil0f%2522%252C%2522activeTabId%2522%253A%2522clw0p8510004c3p6gxo3ivu91%2522%257D%252C%2522clw0p75sm00063p6gbgvq5cit%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clw0p75sm00053p6gg3vy2lku%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252Fform%2522%257D%255D%252C%2522id%2522%253A%2522clw0p75sm00063p6gbgvq5cit%2522%252C%2522activeTabId%2522%253A%2522clw0p75sm00053p6gg3vy2lku%2522%257D%252C%2522clw0p75sm00043p6ggf1568jt%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clw0p75sm00043p6ggf1568jt%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)]

1. you mapped children of `tabs` to `tabs`, I created a new `tab` so it's wrapped as a new field.
2. your autocomplete component is wrong - you store a whole object there, but you need to store `data.value`

```jsx
onChange={(_, data) => input.onChange(data.value)}
```

**Reply by adityapatil9191** [reply]
@rvsia thanks man it works !!! I knew I was missing something.

**Reply by adityapatil9191** [reply]
@rvsia in this example, how can I handle the functionality of keeping only one accordion open at a time. I tried passing index as a property to the accordion tab component but that did not work.

**Reply by rvsia** [reply]
Hey, you need to write your own component to do that as usual in React.


Example:

```jsx
Accordion = ({ fields }) => {
   const [selected, setSelected] = useState()

    return fields.map(tab => 
                <Tab isOpen={selected === tab.name} onSelected={() => setSelected(tab.name)} }>
                        {renderForm(tab.fields)}
                 </Tab>
   )
}
```

**Reply by adityapatil9191** [reply]
@rvsia thanks that is precisely what I am trying . Have been trying to resolve the problem for quite sometime https://codesandbox.io/p/sandbox/react-data-driven-autocomplete-forked-wssh35 . I have updated the accordion MUI component based on the open close logic and state management. This was working fine before. I am not sure if this has got to anything with the tab component though. Please find the reference sandbox url.

**Reply by adityapatil9191** [reply]
This is the exact error that I am getting on my project as well TypeError
TypeError
Cannot read properties of undefined (reading 'map')
    at Object.renderForm (https://wssh35.csb.app/node_modules/
data-driven-forms/react-form-renderer/esm/form-renderer/render-form.js:137:17
eval
/src/components/componentMapper.jsx:169:21
  166 |       </Box>
  167 |     </AccordionSummary>
  168 |     <AccordionDetails>
> 169 |       {formOptions.renderForm(field.fields)}
      |                   ^
  170 |     </AccordionDetails>
  171 |   </Accordion>
  172 | ));
View compiled

**Reply by rvsia** [reply]
You need to extract the logic outside individual tab components one level up to the parent [sandbox](https://codesandbox.io/p/sandbox/react-data-driven-autocomplete-forked-56m5d9)

**Reply by adityapatil9191** [reply]
Thanks buddy.  I am new to this library so learning it . Thanks again

**Comment by adityapatil9191** [comment]
Hello Richard,

I have added a comment here. Could you please look into it and help me ?

Thanks,
Aditya
